Update hwdec_cuda to work with libplacebo ra_pl contexts

philipl opened the pull request:
This patch series updates hwdec_cuda to work with the libplacebo branch. It also includes the switch over to use image (rather than buffer) based interop, which has finally been fixed in the nvidia driver.

philipl committed:
vo_gpu: hwdec_cuda: Implement interop for placebo

This change updates the vulkan interop code to work with the
libplacebo based ra_vk, but also introduces direct VkImage
sharing to avoid the use of the intermediate buffer.

It is also necessary and desirable to introduce explicit
semaphore bsed synchronisation for operations on the shared
images.

Synchronisation means we can safely reuse the same VkImage for every
mapped frame, by ensuring the frame is copied to the VkImage before
mapping the next frame.

This functionality requires a 417.xx or newer nvidia driver, due to
bugs in the VkImage interop in the earlier 411 and 415 drivers.

It's definitely worth the effort, as the raw throughput is about
twice that of implementation using an intermediate buffer.

philipl committed:
vo_gpu: ra_pl: Add getter for pl_gpu

We need access to the underlying pl_gpu to make libplacebo calls
from hwdecs.

philipl committed:
vo_gpu: vulkan: Add back context_win for libplacebo

Feature parity with the original ra_vk obviously requires win32 support,
so let's put it back in.

philipl committed:
vo/gpu: hwdec_cuda: Reorganise backend-specific code

This tries to tidy up the GL vs Vulkan code to be a bit cleaner
and easier to read.

haasn commented on video/out/vulkan/context_win.c:
```diff
@@ -0,0 +1,94 @@
+
+static int win_control(struct ra_ctx *ctx, int *events, int request, void *arg)
+{
+    return vo_w32_control(ctx->vo, events, request, arg);
```
As a result of [ad242535c](https://code.videolan.org/videolan/libplacebo/commit/ad242535cb675250e5776ccccc1e3842386e153e), we can (and most likely, should) implement the `resize` event again now, bringing this file closer to what it looks like in current mpv master. See the current [context_xlib.c](https://github.com/haasn/mpv/blob/libplacebo/video/out/vulkan/context_xlib.c) for an example.

haasn commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -300,105 +288,193 @@ static void cuda_uninit(struct ra_hwdec *hw)
+        .sampleable = true,
+        .sample_mode = format->linear_filter ? PL_TEX_SAMPLE_LINEAR
+                                             : PL_TEX_SAMPLE_NEAREST,
+        .export_handle = handle_type,
```
Missing check to see if this specific `handle_type` is supported. (It's UB to misuse the `pl_gpu` API)

I would probably move this handle decision logic, and the corresponding checks for all relevant objects, to the init function. (Where you currently just check to see if the caps are non-NULL)

haasn commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -300,105 +288,193 @@ static void cuda_uninit(struct ra_hwdec *hw)
+        .size = mp_image_plane_h(&p->layout, n) *
+                mp_image_plane_w(&p->layout, n) *
+                mapper->tex[n]->params.format->pixel_size,
+        .handle_type = handle_type,
```
Again, missing check

haasn commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -300,105 +288,193 @@ static void cuda_uninit(struct ra_hwdec *hw)
+        goto error;
+    }
+
+    evk->sync = pl_sync_create(gpu, handle_type);
```
Missing check

haasn commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -300,105 +288,193 @@ static void cuda_uninit(struct ra_hwdec *hw)
+        if (evk->buf > 0) {
+            CHECK_CU(cu->cuMemFree(evk->buf));
+        }
+        if (evk->plbuf) {
```
These non-NULL checks are technically unneeded, `pl_*_destroy` is safe to call on NULL.

haasn commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -334,24 +333,9 @@ static bool cuda_ext_vk_init(struct ra_hwdec_mapper *mapper,
     }
     mapper->tex[n] = ratex;
```
A few lines above this, when you create the `pl_tex`, you still set `host_writable` even though you no longer need it (since you no longer call `pl_tex_upload`).

haasn commented on wscript:
```diff
@@ -852,7 +852,7 @@ hwaccel_features = [
         'name': 'ffnvcodec',
         'desc': 'CUDA Headers and dynamic loader',
-        'func': check_pkg_config('ffnvcodec >= 8.2.15.3'),
+        'func': check_pkg_config('ffnvcodec >= 8.2.15.7'),
```
Should probably squash this in with the commits that actually need it.

haasn commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -103,9 +115,123 @@ static int check_cu(struct ra_hwdec *hw, CUresult err, const char *func)
+static bool cuda_ext_vk_init(struct ra_hwdec_mapper *mapper,
+                             const struct ra_format *format, int n);
+static void cuda_ext_gl_uninit(struct ra_hwdec_mapper *mapper, int n);
+static void cuda_ext_vk_uninit(struct ra_hwdec_mapper *mapper, int n);
```
At this point you can probably move the implementation of these functions to separate (conditionally compiled) files, and make a single unifying API, like

```c
struct cuda_ext {
    bool (*init)(...);
    void (*uninit)(....);
}
```

And replace the `is_gl` and `is_vk` bools by a `const struct cuda_ext *` pointer. That way we get rid of basically all ifdef in the code and make the platform-specific code a bit clearer.

philipl replied:
I actually started off trying to do this, but it became a mess with the handling of the 'priv' structs, and trying to handle all of the vulkan vs gl logic. I can try and do it again, but I'll handle it as a separate PR.

haasn reviewed (changes requested):
Looks good overall. I would squash a few commits here (specifically the `pl_tex` code with the `pl_buf` code, so we don't introduce logic only do delete it in the next commit)

haasn commented on video/out/opengl/hwdec_cuda.c:
```diff
@@ -300,105 +288,193 @@ static void cuda_uninit(struct ra_hwdec *hw)
+        .w = mp_image_plane_w(&p->layout, n),
+        .h = mp_image_plane_h(&p->layout, n),
+        .d = 0,
+        .format = format->priv,
```
Instead of accessing `format->priv` directly, can you add a "backdoor" function like `static inline const struct pl_fmt *ra_pl_fmt_get(const struct ra_format *format)` to `ra_pl.h`? That way we have a single place to update if we ever make `fmt->priv` point at something other than `pl_fmt`.

philipl committed:
vo/gpu: ra_pl: Add helper to get pl_fmt from ra_format

When interacting directly with libplacebo, we may need to pass a
pl_fmt based on an ra_format. Although the mapping is currently
trivial, it's worth wrapping to make it easy to adapt if this
changes in the future.
